src/Splat.Microsoft.Extensions.DependencyInjection/MicrosoftDependencyResolver.cs

```diff
@@ -0,0 +1,355 @@
+            return service != null;
+        }
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Disposes of the instance.
+        /// </summary>
+        /// <param name="disposing">Whether or not the instance is disposing.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+        }
+
+        private static Type GetDictionaryType(Type serviceType) => _dictionaryType.MakeGenericType(serviceType);
+
+        private void RemoveContractService(Type serviceType)
+        {
+            var dicType = GetDictionaryType(serviceType);
+            var sd = _serviceCollection.SingleOrDefault(s => s.ServiceType == serviceType);
+            _serviceCollection.Remove(sd);
+        }
+
+        private ContractDictionary GetContractDictionary(Type serviceType, bool createIfNotExists)
+        {
+            var dicType = GetDictionaryType(serviceType);
+
+            if (_isImmutable)
+            {
+                return (ContractDictionary)ServiceProvider.GetService(dicType);
+            }
+
+            var dic = _serviceCollection
+                .Where(sd => sd.ServiceType == dicType)
+                .Select(sd => sd.ImplementationInstance)
+                .Cast<ContractDictionary>()
+                .SingleOrDefault();
+
+            if (createIfNotExists && dic == null)
+            {
+                dic = (ContractDictionary)Activator.CreateInstance(dicType);
+                _serviceCollection.AddSingleton(dicType, dic);
+            }
+
+            return dic;
+        }
+
+        private class ContractDictionary
+        {
+            private readonly ConcurrentDictionary<string, IList<Func<object>>> _dictionary = new ConcurrentDictionary<string, IList<Func<object>>>();
+
+            public bool IsEmpty => _dictionary.Count == 0;
+
+            public bool TryRemoveContract(string contract) =>
+                _dictionary.TryRemove(contract, out var _);
+
+            public Func<object> GetFactorý(string contract)
+            {
+                return GetFactories(contract)
+                    .LastOrDefault();
+            }
+
+            public IEnumerable<Func<object>> GetFactories(string contract)
+            {
+                if (_dictionary.TryGetValue(contract, out var collection))
+                {
+                    return collection;
+                }
+
+                return Enumerable.Empty<Func<object>>();
+            }
+
+            public void AddFactory(string contract, Func<object> factory)
+            {
+                var collection = _dictionary.GetOrAdd(contract, _ => new List<Func<object>>());
+                collection.Add(factory);
+            }
+
+            public void RemoveLastFactory(string contract)
+            {
+                if (_dictionary.TryGetValue(contract, out var collection))
```

Comment:
I'm pretty confident this wouldn't be thread safe btw. What happens if users try to access multiple contracts at the same time.

You might be able to use 

https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.addorupdate?view=netframework-4.8#System_Collections_Concurrent_ConcurrentDictionary_2_AddOrUpdate__0__1_System_Func__0__1__1__

So something like

```
_dictionary.AddOrUpdate(contract, default, (key, list) => 
   {
       var lastIndex = list.Count - 1;
                    if (lastIndex > 0)
                    {
                        list.RemoveAt(lastIndex);
                    }

   return list;
   });
```

That should make it a atomic operation.

Reply:
I would be careful with the "add" parameter to be a List<Func<object>>

The issue with the current remote functionality since it's two independent method calls could be an issue. Either we decided to leave a empty entry for that dictionary (easier option), or we have to introduce some sort of lock (and we might be as well make it a Dictionary/List combination then).

Reply:
But we want to remove this contract

Reply:
Well, yeah that's what I mean in the comment above, if the list we return is empty (or null), we can assume there is nothing registered.

Reply:
so the suggestion I have will clear it from the list.

Also be worth just thinking about it doing a if (list.Count == 0) return null; check in there

Reply:
Why assume and not just dump it if it's there for no reason

Reply:
You sure you're not talking about the `AddFactory` method?

https://github.com/reactiveui/splat/blob/c7f5d5b8b613538e412e525472cbdcaf2a5c0ad6/src/Splat.Microsoft.Extensions.DependencyInjection/MicrosoftDependencyResolver.cs#L318-L322

Reply:
I think I get it, lemme check again

Reply:
yes, the above code I suggested, removes items from the list. It does not remove the "contract" entry though since doing that would require a separate operation which would remove the atomic operation nature of it.

It's `AddOrUpdate` and we are mostly interested in the `Update` portion of that method. 

We could be left with a scenario where we have a contract with a empty list and we can deal with that elsewhere.

Reply:
what I mean by with atomic btw, is the Func in the third parameter has to be completed before other operations can be done on the dictionary.

Reply:
Well ConcurrentDictionary is smarter than that, it tends to do locking based on the keys.